staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator.go

```diff
@@ -0,0 +1,149 @@
+	// Options define the requested streams (e.g. stdin, stdout).
+	Options Options
+}
+
+// NewStreamTranslatorHandler creates a new proxy handler. Responder is required for returning
+// errors to the caller.
+func NewStreamTranslatorHandler(location *url.URL, transport http.RoundTripper, responder apimachineryproxy.ErrorResponder, opts Options) *StreamTranslatorHandler {
+	return &StreamTranslatorHandler{
+		Location:  location,
+		Transport: transport,
+		Responder: responder,
+		Options:   opts,
+	}
+}
+
+func (h *StreamTranslatorHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// Create WebSocket server, including particular streams requested.
+	websocketStreams, err := webSocketServerStreams(req, w, h.Options)
+	if err != nil {
+		klog.Errorf("websocket createStreams error: %v", err)
+		h.Responder.Error(w, req, err)
+		return
+	}
+	defer websocketStreams.conn.Close()
+
+	// Create the SPDY client connection proxied through kubelet.
+	tlsConfig, err := utilnet.TLSClientConfig(h.Transport)
+	if tlsConfig != nil && err == nil {
+		// Manually set NextProtos to "http/1.1", since http/2.0 will NOT upgrade a connection.
+		tlsConfig.NextProtos = []string{"http/1.1"}
+	} else {
+		klog.Infof("Unable to unwrap transport %T to get at TLS config: %v", h.Transport, err)
+	}
+	spdyRoundTripper := spdy.NewRoundTripper(tlsConfig)
+	spdyExecutor, err := remotecommand.NewSPDYExecutorForTransports(spdyRoundTripper, spdyRoundTripper, "POST", h.Location)
+	if err != nil {
+		klog.Errorf("websocket createStreams error: %v", err)
+		h.Responder.Error(w, req, err)
+		return
+	}
+
+	// Wire the WebSocket server streams output to the SPDY client input.
+	opts := remotecommand.StreamOptions{}
+	if h.Options.Stdin {
+		opts.Stdin = websocketStreams.stdinStream
+	}
+	if h.Options.Stdout {
+		opts.Stdout = websocketStreams.stdoutStream
+	}
+	if h.Options.Stderr {
+		opts.Stderr = websocketStreams.stderrStream
+	}
+	if h.Options.Tty {
+		opts.Tty = true
+		opts.TerminalSizeQueue = &translatorSizeQueue{resizeChan: websocketStreams.resizeChan}
+	}
+	// Start the SPDY client with connected streams. Output from the WebSocket server
+	// streams will be forwarded into the SPDY client.
+	err = spdyExecutor.StreamWithContext(req.Context(), opts)
+	if err != nil {
+		klog.Errorf("spdyExecutor StreamWithContext() error: %v", err)
+		// Report error back to the WebSocket client.
+		//nolint:errcheck
+		if exitErr, ok := err.(exec.CodeExitError); ok && exitErr.Exited() {
+			rc := exitErr.ExitStatus()
+			websocketStreams.writeStatus(&apierrors.StatusError{ErrStatus: metav1.Status{
+				Status: metav1.StatusFailure,
+				Reason: constants.NonZeroExitCodeReason,
+				Details: &metav1.StatusDetails{
+					Causes: []metav1.StatusCause{
+						{
+							Type:    constants.ExitCodeCauseType,
+							Message: fmt.Sprintf("%d", rc),
+						},
+					},
+				},
+				Message: fmt.Sprintf("command terminated with non-zero exit code: %v", exitErr),
```

Comment:
is this mimicking an existing error response we already get from the kubelet / CRI?

Reply:
it looks like we're narrowly copying exactly one type of response... shouldn't this be transforming *any* structured error?

Reply:
We can get fancier with how we write SPDY execution errors onto the WebSocket error stream. Currently, we

1. (fancier) Transform the returned SPDY errors where the container process exited with a non-zero exit code, and write it to the WebSocket error stream. This ensures we fulfill the spirit of the `v4` error stream, which was all about surfacing the container process exit code.
2. Otherwise, wrap other SPDY execution errors as structured 500-level errors, and write them to the WebSocket error stream.

(structured success status messages are written to the error stream if the previous two cases do not occur).

Please let me know where you believe we should extend the error processing within the `StreamTranslator`